### PR TITLE
[Merged by Bors] - feat: port Data.Setoid.Basic

### PR DESCRIPTION
**Pull request by dupuisf**

### CODE REVIEW

File: Mathlib/Data/Setoid/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+    ext' fun x y =>
+      ⟨fun h => let ⟨a, b, hx, hy, H⟩ := h; hx ▸ hy ▸ H,
+        Quotient.inductionOn₂ x y fun w z h => ⟨w, z, rfl, rfl, h⟩⟩
+  map_rel_iff' := -- fun s t =>
+    ⟨fun h x y hs =>
+      let ⟨a, b, hx, hy, ht⟩ := h ⟨x, y, rfl, rfl, hs⟩
+      t.1.trans' (t.1.symm' <| t.2 <| eq_rel.1 hx) <| t.1.trans' ht <| t.2 <| eq_rel.1 hy,
+      fun h x y hs =>
+      let ⟨a, b, hx, hy, Hs⟩ := hs
+      ⟨a, b, hx, hy, h Hs⟩⟩
```
Comment (dupuisf) [inline Mathlib/Data/Setoid/Basic.lean:465]: This is the last sorry in the file, and I'm a bit stumped here -- if anyone wants to have a crack at it, go for it!

File: Mathlib/Data/Setoid/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+theorems for quotients of arbitrary types.
+
+## Implementation notes
+
+The function `Rel` and lemmas ending in ' make it easier to talk about different
+equivalence relations on the same type.
+
+The complete lattice instance for equivalence relations could have been defined by lifting
+the Galois insertion of equivalence relations on α into binary relations on α, and then using
+`complete_lattice.copy` to define a complete lattice instance with more appropriate
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Basic.lean:28]:
```suggestion
`CompleteLattice.copy` to define a complete lattice instance with more appropriate
```

File: Mathlib/Data/Setoid/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    refine ⟨?_, ?_⟩
+    · intro h x y hs
+      let ⟨a, b, hx, hy, ht⟩ := h ⟨x, y, rfl, rfl, hs⟩
+      exact t.1.trans' (t.1.symm' <| t.2 <| eq_rel.1 hx) <| t.1.trans' ht <| t.2 <| eq_rel.1 hy
+    · intro h x y hs
+      let ⟨a, b, hx, hy, Hs⟩ := hs
+      exact ⟨a, b, hx, hy, h Hs⟩
+
+
+
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Basic.lean:466]:
```suggestion
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**